eeps/eep-0065.md

```diff
@@ -0,0 +1,282 @@
+----------------
+
+In the following text double angle quotation marks are used to
+highlight source code characters.  For example: «`.`» means
+the dot character (full stop).
+
+### Problems with simple prefixes
+
+All of these simple prefixes has to become separate tokens in the scanner
+(lexer): «`bf"`» would constitute the start token for a binary string
+with interpolation syntax.  So would «`bf"""`», «`b"`», «`b"""`»,
+and so on.
+
+The scanner would have to know of all combinations of prefix characters
+and emit distinct tokens for every combination.
+
+Today, the character sequence «`b`», «`f`», «`"`» is scanned as a token
+for the atom «`bf`» followed by the string start token «`"`».
+That combination fails in the parser so it is syntactically invalid today,
+which is what makes simple prefixes a possible language extension.
+
+A simple prefix approach would have to scan a number of characters
+ahead to distinguish between an atom followed by string start
+vs. prefixed string start, and it would be a different number
+of characters depending on which atom characters that have
+been found so far.  This is rather messy.
+
+Furthermore, it is likely that we want the feature of choosing
+[String Delimiters][], especially for regular expressions such as:
+
+    re(^"+.*/.*$)
+
+Among the desired delimiters are «`/`» and «`<`»+«`>`».  The currently
+valid code «`b<X`» meaning atom «`b`» less than «`X`», would instead
+have to be interpreted as prefixed string start «`b<`» with «`X`»
+being the first string content character.
+
+For the «`/`» character we run into similar problems with for example
+«`b/X`», which would be a run-time error today, but if we also would
+want capital letter prefixes, then «`B/X`» is perfectly valid today
+but would become a string start.
+
+There are more likely problems with simple string prefixes:
+«`#bf{`» is today the start of a record named «`bf`», and is
+scanned as punctuation character «`#`», atom «`bf`» and separator «`{`»,
+which the parser sorts out to be a record start.
+
+With simple prefix characters the scanner would have to be rewritten
+to recognize «`#bf`» as a new record token, a rewrite that might cause
+unexpected changes in record handling.  For example, today, «`# bf {`»
+is also a valid record start, so to be completely compatible the scanner
+would have to allow white-space or even newlines within the new record
+token, between «`#`» and the atom characters, which would be really ugly...
+
+For reasons, maybe some of them above, Elixir has chosen to use
+the «`~`» character as the start of a string prefix which they
+call a ["Sigil"][1].
+
+### Sigil
+
+In a general sense, a [Sigil][3], is a prefix to a variable
+that indicates its type, such as «`$I`» in Basic or Perl,
+ where «`$`» is the sigil and «`I`» is the variable.
+
+Here we define a Sigil as a prefix to a string literal
+that indicates how to interpret the string literal.
+It consists of:
+
+1. The character «`~`»
```

Comment:
The other reason I do not like these simple prefixes is because the implications they have on the structure of the language, in that how they affect the lexer and the parser.

I will write a section on my woes about the lexer and the parser, in the EEP.

Today, `bf"my regex"` is scanned (by the lexer) as atom:`bf`, string:`my regex`.  The lexer knows after scanning `b` that it is scanning an atom, and at `"` the atom ends since a new token, a string, starts.

With a simple prefix the lexer has to know which atom start sequences that may be string prefixes, or has to know that if an atom (or variable) ends with `"`, `'` (or `[`) it is a string prefix, not an atom (nor a variable).  This also introduces a small discrepancy between the first and the other characters in the prefix since `0`-`9` and `@` are not allowed as the first character.

I think it would be a really ugly odd bird if how a token ends changes the token type...

Can we leave this for the parser to figure out?

Then the parser would see that atom+string or variable+string should be a string prefix.  This opens for using a quoted_atom+string, which allows a much larger character set in the atom, and looks really odd: `'Bää' "Content"`.  The combination is a parse error today, and who knows which programmer mistakes we open for by allowing it.

But the biggest problem would be that the lexer did not know how to scan the string with respect to e.g escape sequences.  So either the lexer would have to peek at the atoms and variables it scans, and keep that in its state for a string that might follow, which would be doing the job of the parser, or the parser would have to translate escape sequences in the string, which is the job of the lexer.  ... The lexer would also have to know that after an atom or a variable, a quoted atom is not an atom but a string, and possibly a `[` is not a punctuation character, but instead the start of a string.  Again, the parser's job.

Reply:
> I think it would be a really ugly odd bird if how a token ends changes the token type...

I read this sentence and I thought: well, don't you already change the token based on the presence of the `@` character? And then I tried it out in the shell:

```erlang
1> foo@bar(bar).
** exception error: undefined shell command foo@bar/1
```

And of course `foo@bar` is a valid function name, which is simple and it makes total sense! In other words, I understand it better now, thank you.

> But the biggest problem would be that the lexer did not know how to scan the string with respect to e.g escape sequences. 

Just to close the loop on this conversation, depending if you want to allow custom sigils or to make them extensible, that's a decision you will have to make anyway. That's because the sigil is the one that decides the escape characters. After all, the `r` sigil uses different escape sequences than `s`.

So there is an option of saying that the sigil goes as a raw string all the way up to the Erlang Abstract Format, and the compiler/parse transform escapes and interpolates the sigil contents as they prefer. This would give you the maximum extensibility.

Elixir leaves the encoding up to the sigil (and [we provide helpers functions](https://hexdocs.pm/elixir/Macro.html#unescape_string/1)) but interpolation is made part of the tokenizer. Erlang is much simpler to tokenize/parse, so you could leave both of them up to the sigil author, if you want to.

Reply:
> After all, the `r` sigil uses different escape sequences than s.

Oh, does it?  Interesting! What are the differences / where is that documented?

And, yes, we need to decide on the matter of extensible sigils, and then how the lexer and the parser should handle them...

Reply:
Here is an example: `\s` in a string is a whitespace. `\s` in a regex means any whitespace character (\n, \t, \s). Not only that, when you see `\[` in a string, you silently discard the `\`. However, in a regex, you want to keep the escape character around and send it all the way down to the regex engine.

Here is the unescape map for regexes from Elixir:

https://github.com/elixir-lang/elixir/blob/bc3c915563a59c40c476fd2d512d43ccdac793e6/lib/elixir/lib/regex.ex#L857

Here is the default unescape map for strings:

https://github.com/elixir-lang/elixir/blob/bc3c915563a59c40c476fd2d512d43ccdac793e6/lib/elixir/src/elixir_interpolation.erl#L251-L265

Returning false means "leave the escape character in, do not remove it".